### DISCUSSION TITLE
Write out a string to a serial USB device before loading kernel and initrd?

### DISCUSSION THREAD

**Discussion by AndreasUfert** [ipxe/ipxe#934] [Q&A]
Hi,

in a PXE setup using iPXE I want to switch off a hardware watchdog very early in the boot process (i.e. before loading kernel and initrd) by writing some magic string to a serial USB device (in this case: 0403:6015 or 03eb:2307 respectively). Is it possible to write a simple iPXE module that opens the device and writes the magic string? Is there already a driver in iPXE that supports these two USB serial adapters (Future Technology and Atmel)?

Regards, Andreas

**Comment by NiKiZe** [comment]
There is no serial usb device support, but it sounds like you want a bit of special driver that does just this.

In EFI mode there could be a firmware driver, but still no way in iPXE to do what you want, but you could chain to external efi code for this.

**Comment by AndreasUfert** [comment]
Thanks for your quick answer. I was kind of fearing that it wouldn't be so easy to achieve as this exceeds my coding skills by far being more of an administrator guy.

Thank you anyway!

Regards, Andreas

**Comment by mcb30** [comment]
@AndreasUfert There is no specific support for this feature.  It could be worth adding USB serial console support to iPXE: the console wouldn't be usable until after device probe had completed (and so couldn't catch a large class of debug messages), but it could still be useful.

One question: why do you want to disable the watchdog before starting the kernel?  The normal idea of a hardware watchdog is that it keeps running, and the loaded operating system is supposed to write to it periodically to indicate that the OS is still alive.

**Comment by AndreasUfert** [comment]
@mcb30 In production, all my devices boot from local disk and with a watchdog timeout of about 60 seconds everything just works: The watchdog trigger process in the OS is writing to the watchdog periodically, all is fine.

However, there are maintenance scenarios in which all devices in a network are booted from PXE. They all are powered on simultaneously (power cycling the entire landscape is the only option) and simultaneously read kernel and rootfs (presently via TFTP, soon via HTTP). The PXE server is a single service laptop and serving all files for all devices through one interface in parallel takes time.

Since the watchdog is activated the moment the device is powered on the only option is setting the hardware watchdog timeout to much longer than 60 seconds, i.e. a couple of minutes.

I was hoping I could switch of the watchdog very early in the PXE process (i.e. in iPXE) so transferring and booting kernel and rootfs could take as long as it takes.

**Reply by mcb30** [reply]
Thanks for the explanation.  iPXE does have some support for the UEFI abstract watchdog timer, which could be generalised to allow for different implementations such as a USB watchdog.

Do you have an Amazon or similar product link for the watchdog timer that you are using?

**Reply by AndreasUfert** [reply]
No, I don't have any product link. It's a customer specific device implementing a watchdog and multiple other functionalities like controlling adaptive display brightness and temperature control.

What I do know is that either 0403:6015 or 03eb:2307 need to be supported and what particular strings I have to write to these serial devices to control the watchdog.
